docs/ACTION.md

```diff
@@ -0,0 +1,143 @@
+
+<a name="actionloop">
+
+### The Action Loop Protocol
+
+The protocol can be specified informally as follow.
+
+- Read one line from standard input (file descriptor 0).
+- Parse the line as a JSON object. Currently the object will be in currently in the format:
+
+```
+{
+ "value": JSON,
+ "namespace": String,
+ "action_name": String,
+ "api_host": String,
+ "api_key": String,
+ "activation_id": String,
+ "deadline": Number
+}
+```
+
+Note however that more values could be provided in future.
+Usually this JSON is read and the values are stored in environment variables, converted to upper case the key and  and adding the prefix `__OW_`.
+
+- The payload of the request is stored in the key `value`. The action should read the field `value` assuming it is a JSON object (note, not an array, nor a string or number) and parse it.
+- The action can now perform its tasks as appropriate. The action can produce log writing  in standard output (file descriptor 1) and standard error (file descriptor 3) . Note that those corresponds to file descriptors 1 and 2.
+- The action will receive also file descriptor 3 for returning results. The result of the action must be a single line (without embedding newlines - newlines in strings must be quoted) written in file descriptor 3.
+- The action should not exit now, but continue the loop, reading the next line and processing as described before, continuing forever.
+
+### Using shell scripts
+
+The `actionloop` image works actually with executable in Linux sense, so also scripts are acceptable.
+
+In the current actionloop image there is `bash` and the `jq` command, so you can for example implement the actionloop with a shell script like this:
+
+```bash
+#!/bin/bash
+# read input forever line by line
+while read line
+do
+   # parse the in input with `jq`
+   name="$(echo $line | jq -r .name.value)"
+   # log in stdout
+   echo msg="hello $name"
+   # produce the result - note the fd3
+   echo '{"hello": "'$name'"}' >&3
+done
+```
+
+Note here we are just interested in the payload, but in general you may also want to retrieve other fields.
+
+Note the `actionloop` image will accept any source and will try to run it (if it is possible), while the `actionloop-golang-v1.11`  will instead try to compile the sources assuming it is Golang instead.
+
+<a name="actionloopgo">
+
+### Providing your own ActionLoop implementation
+
+By default the runtime expects you provide a main function that will serve one request, and will add a default implementation of the ActionLoop protocol when compiling.
+
+You can however overwrite the default protocol and provide your how implementation of the ActionLoop. If you do so, you will have to take care of opening file descriptors, reading input, parse JSON and set environment variables.
+
+To overwrite the default ActionLoop you can do this either sending a single file source actino or a zip action.
```

Comment:
To overwrite the default ActionLoop you can do this either sending a single file source action or a zip action.
```suggestion
To overwrite the default ActionLoop you can do this either sending a single file source action or a zip action.
```